src/cloudant/index.py

```diff
@@ -0,0 +1,187 @@
+                raise ValueError("Invalid argument: {0}".format(k))
+        extras = python_to_couch(self.options)
+        return extras
+
+    def __getitem__(self, key):
+        """
+        implement key access and slicing.
+
+        key can be either a single value as a string or list which will be
+        passed as the key to the query for entries matching that key or
+        a slice object.
+
+        Slices with integers will be interpreted as skip:limit style pairs,
+        eg with [100:200] meaning skip 100, get 200 records.
+
+        Slices with strings/lists will be interpreted as startkey/endkey
+        style keys.
+
+        """
+        extras = self._prepare_extras()
+        if isinstance(key, basestring):
+            data = self._ref(key=key, **extras)
+            return data['rows']
+
+        if isinstance(key, list):
+            data = self._ref(key=key, **extras)
+            return data['rows']
+
+        if isinstance(key, slice):
+            # slice is startkey and endkey if str or array
+            str_or_none_start = isinstance(
+                key.start, (basestring, list)
+                ) or key.start is None
+            str_or_none_stop = isinstance(
+                key.stop, (basestring, list)
+                ) or key.stop is None
+            if str_or_none_start and str_or_none_stop:
+                # startkey/endkey
+                if key.start is not None and key.stop is not None:
+                    data = self._ref(
+                        startkey=key.start,
+                        endkey=key.stop,
+                        **extras
+                    )
+                if key.start is not None and key.stop is None:
+                    data = self._ref(startkey=key.start, **extras)
+                if key.start is None and key.stop is not None:
+                    data = self._ref(endkey=key.stop, **extras)
+                if key.start is None and key.stop is None:
+                    data = self._ref(**extras)
+                return data['rows']
+            # slice is skip:limit if ints
+            int_or_none_start = isinstance(
+                key.start, (int)
+                ) or key.start is None
+            int_or_none_stop = isinstance(
+                key.stop, (int)
+                ) or key.stop is None
+            if int_or_none_start and int_or_none_stop:
+                if key.start is not None and key.stop is not None:
+                    data = self._ref(
+                        skip=key.start,
+                        limit=key.stop,
+                        **extras
+                    )
+                if key.start is not None and key.stop is None:
+                    data = self._ref(skip=key.start, **extras)
+                if key.start is None and key.stop is not None:
+                    data = self._ref(limit=key.stop, **extras)
+                # both None case handled above
+                return data['rows']
```

Comment:
switch statement?  :) 

I agree that it could use some cleanup, I do expect it to evolve a bit more though.